(xmlsec-openssl) Fix AWS-LC support

smittals2 opened the pull request:
This PR explicitly ifdef's xmlsec code with the macro OPENSSL_IS_AWSLC as discussed in issue #908

lsh123 commented on configure.ac:
```diff
@@ -825,7 +825,7 @@ if test "z$OPENSSL_FOUND" = "zyes" ; then
             #include <openssl/crypto.h>
             #if OPENSSL_VERSION_NUMBER >= 0x10100010L
-            #ifdef OPENSSL_IS_BORINGSSL
+            #if defined(OPENSSL_IS_BORINGSSL) || defined(OPENSSL_IS_AWSLC)
```
This change is probably wrong -- this is configure script that detects Boring SSL version compatibility.

Are there any minimum AWS-LC version supported? I would prefer to add a separate check for AWS-LC min version compatibility instead

smittals2 replied:
Ahh good catch, thank you for pointing that out. I've added a separate check for AWS-LC. Had to make some modifications in testrun.sh as well to account for the new AWS-LC version string.

lsh123 commented on tests/testDSig.sh:
```diff
@@ -1837,6 +1782,79 @@ execDSigTest $res_fail \
     "--trusted-$cert_format certs/rsa-ca-cert.$cert_format"
 
+# Skip these tests: AWS-LC will fail verification with MD5 certificate digests
+if [ "z$crypto" = "zopenssl" -a "z$xmlsec_openssl_flavor" != "zawslc" ] ; then
```
This condition is incorrect since it disables these tests not just for AWS-LC but also for all other crypto libraries like NSS

lsh123 commented on tests/testDSig.sh:
```diff
@@ -23,6 +23,23 @@ else
     url_map_rfc3161=""
 fi
 
+# What flavour of OpenSSL do we have?
```
why do you need this here?

https://github.com/lsh123/xmlsec/blob/7809c86c3e1f5fae7dcbd1e7d50dfd94616c6863/tests/testrun.sh#L80

smittals2 replied:
Added it so I can detect AWS-LC and turn off the relevant cert tests below (as is being done in testrun.sh for the ossl-store tests).

lsh123 commented on configure.ac:
```diff
@@ -819,6 +819,26 @@ if test "z$OPENSSL_FOUND" = "zyes" ; then
+            #endif
+        ],[
+            OPENSSL_VERSION="AWSLC >= 1.1.1"
+            enable_ripemd160=no
```
you probably want to disable md5 here too

smittals2 replied:
We support MD5, just not verification of certs that use MD5 for their digest algo (I think this is why the tests use the OPENSSL_ENABLE_MD5_VERIFY macro?). Should we still turn it off here in that case?

lsh123 commented on tests/testDSig.sh:
```diff
@@ -1837,6 +1782,79 @@ execDSigTest $res_fail \
     "rsa x509" \
     "--trusted-$cert_format certs/rsa-ca-cert.$cert_format"
 
+# Skip these tests: AWS-LC will fail verification with MD5 certificate digests
```
I think the better way to do it is to disable MD5 for AWS-LC and then simply check for md5 too:

"sha1 rsa-sha1" ==> "md5 sha1 rsa-sha1"

lsh123 reviewed (changes requested):
Do you mind separating code changes from tests changes? I think there is a better way to do the test changes but it's hard to explain (I tried but I am not sure I did a good job). I can update tests separately.

smittals2 commented:
> Do you mind separating code changes from tests changes? I think there is a better way to do the test changes but it's hard to explain (I tried but I am not sure I did a good job). I can update tests separately.

I'll remove all testing-related changes from this PR. I just want to highlight the two changes that will be needed for AWS-LC compatibility:

1) OSSL-Store Test Disablement: The testrun.sh script needs modification. Currently uses a switch statement to identify OpenSSL variants. I have already added the AWS-LC check in configure.ac as well as setting the version string to AWS-LC. We need to plumb this through into testrun.sh (which this PR does but I have removed for now). 

2) MD5 Certificate Handling: Need a solution for handling MD5 digest certificates in testDSig.sh. Whether that be completely disabling MD5 or using ifdef conditions for specific certificates (I would prefer the latter to still allow MD5 for other use cases if any). 

This PR now contains only code changes. Please let me know if you'd prefer a separate PR to discuss testing changes or another approach to handling the testing modifications.

Looking forward to your feedback.

lsh123 commented:
lgtm; I will let CI run and will merge if everything is green. 

I hear you about tests -- let me think a bit how to do it. It's a bit unusual that MD5 is enabled but certs with MD5 are disabled.

lsh123 commented:
See PR #911 for tests
